### ISSUE TITLE
Layout snippets feature causes command-line recipe execution to silently terminate

### ISSUE THREAD

**Issue by DanielStolt** [OrchardCMS/Orchard#6321]
This happens when:

- One or more snippets exist in a module which is not a theme
- A setup recipe is executed from the command line
- `Orchard.Layout.Snippets` is enabled by the `<Features>` step of the recipe
- The base URL of the site is set in the recipe *before* the content step
- The recipe contains a `<Content>` step
- The `<Content>` step contains at least one content item with a LayoutPart

Currently repros on tip of `1.10`. Steps:

1. Add a sample snippet to Orchard.Blogs\Views named `SampleSnippet.cshtml`.
2. Add a `<Settings>` element and a `<SiteSettingsPart>` to top of the default recipe.
3. Add a `<Content>` step to recipe.
3. Add the `Orchard.Layouts.Snippets` feature to the `<Features>` element.
4. Add a sample `<Page>` element to the recipe.
5. Start `Orchard.exe` from command line.
6. Execute the command `setup /SiteName:"MySite" /DatabaseProvider:SqlCe /AdminUsername:admin /AdminPassword:password /Recipe:"Default"`.
7. Examine the recipe log and `Settings.txt` file. You will find that the site is still in `Initializing` state and that the recipe had silently terminated without any error, before running to completion.

Here's a full recipe that reproduces the issue:

```
<?xml version="1.0"?>
<Orchard>
    
    <Recipe>
        <Name>Default</Name>
        <Description>The default recipe for an Orchard site that includes pages, blogs, custom content types, comments, tags, widgets and basic navigation.</Description>
        <Author>The Orchard Team</Author>
        <WebSite>http://orchardproject.net</WebSite>
        <Category>Default</Category>
        <Tags></Tags>
        <Version>1.0</Version>
        <IsSetupRecipe>true</IsSetupRecipe>
    </Recipe>

    <Feature enable="Orchard.Blogs,Orchard.Comments,Orchard.Tags,Orchard.Alias,Orchard.Autoroute,
                   TinyMce,Orchard.MediaLibrary,Orchard.ContentPicker,Orchard.PublishLater,
                   Orchard.Resources,Orchard.Widgets,Orchard.ContentTypes,
                   Orchard.Scripting,Orchard.Scripting.Lightweight,PackagingServices,Orchard.Packaging,
                   Orchard.Projections,Orchard.Fields,Orchard.OutputCache,Orchard.Taxonomies,Orchard.Workflows,
                   Orchard.Layouts,Orchard.Layouts.Tokens,Orchard.Layouts.Snippets,
                   TheThemeMachine" />

    <Settings>
        <SiteSettingsPart SupportedCultures="en-US;sv-SE" BaseUrl="http://localhost:30321/OrchardLocal" MaxPagedCount="0" MaxPageSize="0" PageSize="10" PageTitleSeparator=" - " SiteCalendar="" SiteCulture="sv-SE" SiteName="My Site" SiteSalt="ca5a5077e2474f30bc65e187e75c043f" SiteTimeZone="W. Europe Standard Time" SuperUser="admin" UseCdn="false" />
    </Settings>

    <ContentDefinition>
        <Types>
            <Page ContentTypeSettings.Draftable="True" TypeIndexing.Indexes="Search">
                <TagsPart />
                <LocalizationPart />
                <TitlePart/>
                <AutoroutePart />
                <MenuPart />
            </Page>
            <BlogPost ContentTypeSettings.Draftable="True" TypeIndexing.Indexes="Search">
                <CommentsPart />
                <TagsPart />
                <LocalizationPart />
                <TitlePart/>
                <AutoroutePart />
            </BlogPost>
        </Types>
        <Parts>
            <BodyPart BodyPartSettings.FlavorDefault="html" />
        </Parts>
    </ContentDefinition>

    <Settings />

    <Migration features="*" />

    <Content>
        <Page Id="/alias=welcome-to-orchard" Status="Published">
            <AutoroutePart Alias="welcome-to-orchard" UseCustomPattern="false" UseCulturePattern="false" PromoteToHomePage="false" />
            <CommonPart Owner="/User.UserName=admin" CreatedUtc="2016-01-31T12:49:19.6178032Z" PublishedUtc="2016-01-31T12:49:19.6178032Z" ModifiedUtc="2016-01-31T12:49:19.5197848Z" />
            <LayoutPart>
                <LayoutData>{"elements":[{"typeName":"Orchard.Layouts.Elements.Canvas","data":"","exportableData":"","index":0,"elements":[{"typeName":"Orchard.Layouts.Elements.Grid","data":"","exportableData":"","index":0,"elements":[{"typeName":"Orchard.Layouts.Elements.Row","data":"","exportableData":"","index":0,"elements":[{"typeName":"Orchard.Layouts.Elements.Column","data":"Width=12","exportableData":"","index":0,"elements":[{"typeName":"Orchard.Layouts.Elements.Html","data":"Content=%3cp%3eYou%27ve+successfully+setup+your+Orchard+Site+and+this+is+the+homepage+of+your+new+site.%0d%0aHere+are+a+few+things+you+can+look+at+to+get+familiar+with+the+application.%0d%0aOnce+you+feel+confident+you+don%27t+need+this+anymore%2c+you+can%0d%0a%3ca+href%3d%22Admin%2fContents%2fEdit%2f9%22%3eremove+it+by+going+into+editing+mode%3c%2fa%3e%0d%0aand+replacing+it+with+whatever+you+want.%3c%2fp%3e%0d%0a%3cp%3eFirst+things+first+-+You%27ll+probably+want+to+%3ca+href%3d%22Admin%2fSettings%22%3emanage+your+settings%3c%2fa%3e%0d%0aand+configure+Orchard+to+your+liking.+After+that%2c+you+can+head+over+to%0d%0a%3ca+href%3d%22Admin%2fThemes%22%3emanage+themes+to+change+or+install+new+themes%3c%2fa%3e%0d%0aand+really+make+it+your+own.+Once+you%27re+happy+with+a+look+and+feel%2c+it%27s+time+for+some+content.%0d%0aYou+can+start+creating+new+custom+content+types+or+start+from+the+built-in+ones+by%0d%0a%3ca+href%3d%22Admin%2fContents%2fCreate%2fPage%22%3eadding+a+page%3c%2fa%3e%2c+or+%3ca+href%3d%22Admin%2fNavigation%22%3emanaging+your+menus.%3c%2fa%3e%3c%2fp%3e%0d%0a%3cp%3eFinally%2c+Orchard+has+been+designed+to+be+extended.+It+comes+with+a+few+built-in%0d%0amodules+such+as+pages+and+blogs+or+themes.+If+you%27re+looking+to+add+additional+functionality%2c%0d%0ayou+can+do+so+by+creating+your+own+module+or+by+installing+one+that+somebody+else+built.%0d%0aModules+are+created+by+other+users+of+Orchard+just+like+you+so+if+you+feel+up+to+it%2c%0d%0a%3ca+href%3d%22http%3a%2f%2forchardproject.net%2fcontribution%22%3eplease+consider+participating%3c%2fa%3e.%3c%2fp%3e%0d%0a%3cp%3eThanks+for+using+Orchard+%e2%80%93+The+Orchard+Team+%3c%2fp%3e","exportableData":"","index":0,"elements":[],"isTemplated":false,"htmlId":null,"htmlClass":null,"htmlStyle":null,"rule":null}],"isTemplated":false,"htmlId":null,"htmlClass":null,"htmlStyle":null,"rule":null}],"isTemplated":false,"htmlId":null,"htmlClass":null,"htmlStyle":null,"rule":null},{"typeName":"Orchard.Layouts.Elements.Row","data":"","exportableData":"","index":1,"elements":[{"typeName":"Orchard.Layouts.Elements.Column","data":"Width=4","exportableData":"","index":0,"elements":[{"typeName":"Orchard.Layouts.Elements.Html","data":"Content=%3ch2%3eFirst+Leader+Aside%3c%2fh2%3e%0d%0a%3cp%3eLorem+ipsum+dolor+sit+amet%2c+consectetur+adipiscing+elit.+Curabitur+a+nibh+ut+tortor+dapibus+vestibulum.%0d%0aAliquam+vel+sem+nibh.+Suspendisse+vel+condimentum+tellus.%3c%2fp%3e","exportableData":"","index":0,"elements":[],"isTemplated":false,"htmlId":null,"htmlClass":null,"htmlStyle":null,"rule":null}],"isTemplated":false,"htmlId":null,"htmlClass":null,"htmlStyle":null,"rule":null},{"typeName":"Orchard.Layouts.Elements.Column","data":"Width=4","exportableData":"","index":1,"elements":[{"typeName":"Orchard.Layouts.Elements.Html","data":"Content=%3ch2%3eSecond+Leader+Aside%3c%2fh2%3e%0d%0a%3cp%3eLorem+ipsum+dolor+sit+amet%2c+consectetur+adipiscing+elit.+Curabitur+a+nibh+ut+tortor+dapibus+vestibulum.%0d%0aAliquam+vel+sem+nibh.+Suspendisse+vel+condimentum+tellus.%3c%2fp%3e","exportableData":"","index":0,"elements":[],"isTemplated":false,"htmlId":null,"htmlClass":null,"htmlStyle":null,"rule":null}],"isTemplated":false,"htmlId":null,"htmlClass":null,"htmlStyle":null,"rule":null},{"typeName":"Orchard.Layouts.Elements.Column","data":"Width=4","exportableData":"","index":2,"elements":[{"typeName":"Orchard.Layouts.Elements.Html","data":"Content=%3ch2%3eThird+Leader+Aside%3c%2fh2%3e%0d%0a%3cp%3eLorem+ipsum+dolor+sit+amet%2c+consectetur+adipiscing+elit.+Curabitur+a+nibh+ut+tortor+dapibus+vestibulum.%0d%0aAliquam+vel+sem+nibh.+Suspendisse+vel+condimentum+tellus.%3c%2fp%3e","exportableData":"","index":0,"elements":[],"isTemplated":false,"htmlId":null,"htmlClass":null,"htmlStyle":null,"rule":null}],"isTemplated":false,"htmlId":null,"htmlClass":null,"htmlStyle":null,"rule":null}],"isTemplated":false,"htmlId":null,"htmlClass":null,"htmlStyle":null,"rule":null}],"isTemplated":false,"htmlId":null,"htmlClass":null,"htmlStyle":null,"rule":null}],"isTemplated":false,"htmlId":null,"htmlClass":null,"htmlStyle":null,"rule":null}]}</LayoutData>
            </LayoutPart>
            <TagsPart Tags="" />
            <TitlePart Title="Welcome to Orchard!" />
        </Page>
    </Content>

    <Command>
        layer create Default /LayerRule:"true" /Description:"The widgets in this layer are displayed on all pages"
        layer create Authenticated /LayerRule:"authenticated" /Description:"The widgets in this layer are displayed when the user is authenticated"
        layer create Anonymous /LayerRule:"not authenticated" /Description:"The widgets in this layer are displayed when the user is anonymous"
        layer create Disabled /LayerRule:"false" /Description:"The widgets in this layer are never displayed"
        layer create TheHomepage /LayerRule:"url '~/'" /Description:"The widgets in this layer are displayed on the home page"
        site setting set baseurl
        menu create /MenuName:"Main Menu"
        menuitem create /MenuPosition:"0" /MenuText:"Home" /Url:"~/" /MenuName:"Main Menu"
        widget create MenuWidget /Title:"Main Menu" /RenderTitle:false /Zone:"Navigation" /Position:"1" /Layer:"Default" /Identity:"MenuWidget1" /MenuName:"Main Menu"
        theme activate "The Theme Machine"
    </Command>
    
</Orchard>
```

**Comment by sfmskywalker** [comment]
We looked into it, and this is what is happening:

1. The SnippetHarvester renders the discovered Snippet shapes.
2. Because this rendering happens outside an HTTP context (since it's Orchard.exe), the `MvcModule.HttpContextPlaceholder` class sets the static `HttpContext.Current` property to a new instance of `HttpContext`, acting as a fake. This is a requirement to support rendering of shapes due to some MVC's HTML helper methods relying on the static `HttpContext.Current` property (which is a real shame; it should use IoC to resolve the current HTTP context, but that's another topic).
3. However, before the SnippetHarvester executes, the `ContextState` class will already have stored the execution ID in the `CallContext`. However, that code performs a check on `HttpContext.Current`. If it is not-null, it assumes a valid HTTP context and therefore stores the information in its `Items` collection. Otherwise, it stores the information in the `CallContext`.

In other words:
1. The execution ID is first stored in CallContext.
2. A shape is being rendered (causing `HttpContext.Current` to be set to a fake instance)
3. The next time the execution ID is requested, `StateContext` tries to read it from `HttpContext.Current.Items` rather than `CallContext`
4. It's a bug

The easiest and most robust way I see to fixing this is by adding another check in the `ContextState.GetState` and `ContextState.SetState` methods. Instead of:

```C#
if (HttpContext.Current == null)
```

The check should be:

```C#
if (HttpContext.Current == null && !HttpContext.Current.Items.Contains(MvcModule.IsBackgroundHttpContextKey)) {
```

Although that isn't pretty, it does make sense: the need to set the static `HttpContext.Current` already is a hack, but just to support rendering of Razor code that relies on certain HTML helpers. So as far as the `ContextState` should be concerned, it should only need to work with real HttpContexts, not fake ones.

Ideally we would be able to work with HttpContext.Current everywhere, but there is probably a good reason we're using CallContext. @jtkech and @sebastienros have probably more to say on this matter.

**Comment by jtkech** [comment]
@sfmskywalker you're totally right.

One reason to use `CallContext.LogicalGet/SetData` is to carry contexts among threads switching due to async code. As done by asp for the `HttpContext.Current` in a normal web request. In fact,  to store `HttpContext.Current`, asp uses `CallContext.Get/SetData` that isn't thread affinative. This is asp that carry the context among threads.

When there is no ambient HttpContext, we first create an  `HttpContextPlaceholder` when a work context scope is created. Then, at this point, at the end of the scope creation we could create a fake HttpContext and update the `HttpContext.Current`, **but it would not be thread affinative as in a normal web request**.

So, i've first opted not to update `HttpContext.Current`. But when we had the issue of view rendering, i've added this update (if needed) when the `HttpContextPlaceholder.Request` property is used.

But, here, because the **`ContextState` is also not thread affinative**, maybe it would be usefull to do this as soon as possible. But for the above reason i thought it was not a good thing to do...

We know that now in some places we have to replace `httpContextBase == null` with `httpContextBase.IsBackGroundContext()`, add some not implemented properties as needed... And we have to **limit the direct usage of `HttpContext.Current`** when it's possible.

But when we are forced to use `HttpContext.Current`, **your solution is the best one**. Note that there is already the static `HttpContextAccessor.IsBackgroundHttpContext(HttpContext httpContext)` method. Do you think it's good to use it if we make it public and / or make an extension?

**But, here, maybe a better solution would be that `ContextState.cs` always uses `CallContext.LogicalGet/SetData`.** And never access `HttpContext.Current`.


Best

**Comment by sfmskywalker** [comment]
Hey @jtkech , thanks for chiming in.

Interesting suggestion to simply stick with `CallContext`. This would simplify the code, and since ASP.NET uses it internally as well anyway, why not?

> Off-topic: Perhaps in **ASP.NET vNext** we could do it the other way around where even the fake background `HttpContext` is not thread affinitive, so we could use that instead of `CallContext`.

Although we shouldn't encourage code to check if the HttpContext is a background / fake one, I think it would be better to reuse the static `HttpContextAccess.IsBackgroundHttpContext() as per your suggestion.

@sebastienros: any thoughts on JT's suggestion regarding the use of `CallContext.LogicalGet/SetData`?

**Comment by jtkech** [comment]
@sfmskywalker, @sebastienros 

It can be done easily and we have to remember to wrap data in an `ObjectHandle` (as done with the workcontext). This because when there are cross domain calls (as through the command line) data need to be serializable. Let me know if i have to do a PR for this and / or to reuse the static `HttpContextAccessor.IsBackgroundHttpContext()`, and in which branch.

Note: with the `Logical CallContext` data are carried not only through an async code flow, but also when explicitly starting parallel tasks. But the last feature is not so usefull, for thread safety most of the time you have to create a new scope for each task.

Just seen [this](https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNetCore.Http/HttpContextAccessor.cs) where asp now do the same for HttpContext through an IHttpContextAccessor. In the dnx version (dnxxxx == netxxx) they use the `Logical CallContext` and wrap data in an `ObjectHandle`! Orchard 2 uses dnxcorexxx (== dotnetxxx), so i think that `AsyncLocal<HttpContext>` will be used in place. Not sure it's the same as the `Logical CallContext` for parallel tasks, but at least it will work with async code.

So, with vnext, maybe we will be able to store a fake HttpContext that will be thread affinative!

That said, i first worked to make workcontext compatible with async code, then i've tried it for `HttpContext.Current`. But as said, because it uses `Get/SetData (not LogicalGet/SetData)`, when not in a normal web request, it is not thread affinative. You can overcome this limitation by storing an object that inherit from `IThreadAffinative`. But you can't do this with the `System.Web.HttpContext sealed class`.

But interesting, i will try again to do something around this when i will have time. Oh, just seen that  `AsyncLocal<>` may be implemented in net 4.6, maybe it will be also used for `HttpContext.Current`!

Best

**Comment by DanielStolt** [comment]
@jtkech I think it makes a lot of sense to simply always use `CallContext` and it would be great if you created a PR for this. Question is to target `dev` on `1.10` - I honestly am not sure. On the one hand it seems like a better version of the fix we just committed to `1.10`, on the other hand it has a much bigger impact since it will change the storage for web hosted scenarios as well. Probably `dev` would be the better place.

**Comment by jtkech** [comment]
@DanielStolt ok, i will do it asap.

Here other infos about some constraints, no worry, just if it can help to always find good compromises.

No luck, with `CallContext Logical Data` it would be so easy to retrieve contexts everywhere if we could use simple data store keys as "WorkContext" and "HttpContext". But, as seen in some rare cases, we could not have nested scopes (or interleaved as seen in an unit test) in the same execution thread.

This could be easily fixed by starting a new thread, then when a new context is written in the logical data store, it will not be shared even by using the same simple key. You can see such a use case when using `Parallel.ForEach()` to activate shell and use processing engine in each task.

But, to keep some existing code, i needed to re-use shell keys for the data store. The result is that you are forced to resolve a wca / hca to get a workcontext / httpcontext (as done in httpContextAccessor).

Finally, in another side, when there is no ambient httpcontext we allow some patterns, as using processing engine in a `Parallel.ForEach()`. This would not be possible in a normal web request because HttpContext.Current properties are not thread safe, anyway it would be lost (maybe not with dnx).

So, same problem if we could always provide a valid HttpContext.Current that could carry different workcontexts. That's why we use the Logical CallContext to carry workcontexts, and each workcontext to carry an HttpContextBase, and in some places we update HttpContext.Current.

Best.
